# Advanced - Pandas: Data Cleaning
___
- EG: `df.dropna()`
- For this section, we will work on a mini Project to learn about Data Cleanup. Specifically, we need to Clean up our Data Frame
- Scenario: We have a co-worker that wants Access to this Data but they have specific Requirements
    - First: We have ~33,000 salary_rate Records. However, the Dataset is ~800,000.
    - This co-worker wants those NaN (Not A Number) Values filled in with something like a Median Value
    - When that's filled in, they want us to go through and Remove any Duplicate Job Entries
        - This is so, they can go through it, look at Jobs and if a Salary isn't there they can see the Median/Expected Salary is for that Job
- **Note**: Everything we do in here, won't be applied to the Final Data Frame that we are working for in the Final Project. 
    - This is more so showing how you can use Functionality to Cleanup Data Sets

In [ ]:
# Imports
import pandas as pd
from datasets import load_dataset
import matplotlib.pyplot as plt

# Loading Data
dataset = load_dataset('lukebarousse/data_jobs')
df = dataset['train'].to_pandas()

# Cleanup
df['job_posted_date'] = pd.to_datetime(df['job_posted_date'])

In [18]:
# Inspecting the Data Frame for the specified Columns
df.loc[:10,'salary_year_avg':'salary_hour_avg']

,salary_year_avg,salary_hour_avg
0,115000.0,45.98
1,115000.0,45.98
2,115000.0,45.98
3,115000.0,45.98
4,115000.0,45.98
5,115000.0,45.98
6,115000.0,45.98
7,115000.0,45.98
8,115000.0,45.98
9,115000.0,45.98


In [10]:
# Get the Median of salary_year_avg
median_salary_year = df['salary_year_avg'].median()

In [11]:
# Get the Median of salary_hour_avg
median_salary_hour = df['salary_hour_avg'].median()

- Now that we have the Median Values figured out(above), we need to move onto the next phase
- Need to figure out how are we going to fill in these NaN Values without Replacing the current Values that are in there
    - Luckily, Pandas has a Method for this!
    - `pandas.DataFrame.fillna`
        - First thing we need to Specify: Value
        - We aren't needing the second Argument for this Method, so we are going to ignore it and move on 
- **Important Note**: Before we start Filtering this down, anytime we go to Modify a Data Frame, you should keep your Original one intact
- Can easily Create a New Data Frame, where you do all your Modifications and changes so the Original one, doesn't get affected and can always refer back to it or not mess up the Source Data
    - Just have to set the New Data Frame equal to the Original Data Frame. (Essentially Creating a Copy to not ruin the Original)

In [17]:
# Create a New Data Frame
df_filled = df

# Specify how we want to fill this in for each of the Columns
# Need to do each Column individually and Run the Method on each one
df_filled['salary_year_avg'] = df_filled['salary_year_avg'].fillna(median_salary_year)
df_filled['salary_hour_avg'] = df_filled['salary_hour_avg'].fillna(median_salary_hour)

In [19]:
# Inspecting the Filled Data Frame for the specified Columns
df_filled.loc[:10,'salary_year_avg':'salary_hour_avg']

,salary_year_avg,salary_hour_avg
0,115000.0,45.98
1,115000.0,45.98
2,115000.0,45.98
3,115000.0,45.98
4,115000.0,45.98
5,115000.0,45.98
6,115000.0,45.98
7,115000.0,45.98
8,115000.0,45.98
9,115000.0,45.98


- Now we can see that it filled in with these Median Values
- First objective that we set out is done! Now moving onto the next Step of Removing the Dupes!
    - And ofc, as you guessed it, there's a Method for that! Specifically for Dropping Duplicates
    - `pandas.DataFrame.drop_duplicates`
    - This Returns a DataFrame with Duplicate Rows Removed
    - Also allows you to Specify Column Lables but we aren't going to specify that just yet
- Again, we are going to Create a New Data Frame to do this so we don't mess up our Source Data and keep the Original intact

In [20]:
# Create New Data Frame and set it Equal to the one with the Replaced NaN Values
df_unique = df_filled

# Remove Duplicates - No need to specify anything inside of it
df_unique.drop_duplicates()

,job_title_short,job_title,job_location,job_via,job_schedule_type,job_work_from_home,search_location,job_posted_date,job_no_degree_mention,job_health_insurance,job_country,salary_rate,salary_year_avg,salary_hour_avg,company_name,job_skills,job_type_skills
0,Senior Data Engineer,Senior Clinical Data Engineer / Principal Clin...,"Watertown, CT",via Work Nearby,Full-time,False,"Texas, United States",2023-06-16 13:44:15,False,False,United States,NaN,115000.0,45.98,Boehringer Ingelheim,NaN,NaN
1,Data Analyst,Data Analyst,"Guadalajara, Jalisco, Mexico",via BeBee México,Full-time,False,Mexico,2023-01-14 13:18:07,False,False,Mexico,NaN,115000.0,45.98,Hewlett Packard Enterprise,"['r', 'python', 'sql', 'nosql', 'power bi', 't...","{'analyst_tools': ['power bi', 'tableau'], 'pr..."
2,Data Engineer,"Data Engineer/Scientist/Analyst, Mid or Senior...","Berlin, Germany",via LinkedIn,Full-time,False,Germany,2023-10-10 13:14:55,False,False,Germany,NaN,115000.0,45.98,ALPHA Augmented Services,"['python', 'sql', 'c#', 'azure', 'airflow', 'd...","{'analyst_tools': ['dax'], 'cloud': ['azure'],..."
3,Data Engineer,LEAD ENGINEER - PRINCIPAL ANALYST - PRINCIPAL ...,"San Antonio, TX",via Diversity.com,Full-time,False,"Texas, United States",2023-07-04 13:01:41,True,False,United States,NaN,115000.0,45.98,Southwest Research Institute,"['python', 'c++', 'java', 'matlab', 'aws', 'te...","{'cloud': ['aws'], 'libraries': ['tensorflow',..."
4,Data Engineer,Data Engineer- Sr Jobs,"Washington, DC",via Clearance Jobs,Full-time,False,Sudan,2023-08-07 14:29:36,False,False,Sudan,NaN,115000.0,45.98,Kristina Daniel,"['bash', 'python', 'oracle', 'aws', 'ansible',...","{'cloud': ['oracle', 'aws'], 'other': ['ansibl..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
785736,Software Engineer,DevOps Engineer,Singapura,melalui Trabajo.org,Pekerjaan tetap,False,Singapore,2023-03-13 06:16:16,False,False,Singapore,NaN,115000.0,45.98,CAREERSTAR INTERNATIONAL PTE. LTD.,"['bash', 'python', 'perl', 'linux', 'unix', 'k...","{'os': ['linux', 'unix'], 'other': ['kubernete..."
785737,Data Analyst,CRM Data Analyst,"Bad Rodach, Jerman",melalui BeBee Deutschland,Pekerjaan tetap,False,Germany,2023-03-12 06:18:18,False,False,Germany,NaN,115000.0,45.98,HABA FAMILYGROUP,"['sas', 'sas', 'sql', 'excel']","{'analyst_tools': ['sas', 'excel'], 'programmi..."
785738,Business Analyst,Commercial Analyst - Start Now,Malaysia,melalui Ricebowl,Pekerjaan tetap,False,Malaysia,2023-03-12 06:32:36,False,False,Malaysia,NaN,115000.0,45.98,Lendlease Corporation,"['powerpoint', 'excel']","{'analyst_tools': ['powerpoint', 'excel']}"
785739,Data Engineer,"Principal Associate, Data Engineer (Remote-Eli...","Newark, New Jersey, Amerika Serikat",melalui Recruit.net,Pekerjaan tetap,False,Sudan,2023-03-12 06:32:15,False,False,Sudan,NaN,115000.0,45.98,Capital One,"['python', 'go', 'nosql', 'sql', 'mongo', 'she...","{'cloud': ['aws', 'snowflake', 'azure', 'redsh..."


- After Running this, we can see we have ~785,000 Rows!
- But, how many Rows did we actually Drop while Running this?
- We can write some Code, to Compare the Length of these different Data Frames

In [22]:
# Create New Data Frame and set it Equal to the one with the Replaced NaN Values
df_unique = df_filled

# Remove Duplicates - No need to specify anything inside of it
df_unique = df_unique.drop_duplicates()

# Get the Length of df_filled
print('Length of Original df:       ', len(df_filled))
# Get the Length of df_unique
print('Length of drop duplicates df:', len(df_unique))
# Subtract the lengths from each other to get the Total Rows Dropped
print('Rows Droppped:               ', len(df_filled)-len(df_unique))

Length of Original df:        785741
Length of drop duplicates df: 785640
Rows Droppped:                101


- By doing this, we now have 101 Rows Removed
- This means, in the Original Dataset of df, there are 101 Duplicate Records still existing (since we didn't run this Cleanup on that Data Frame)
- Also in this Dataset right now, there are 101 Duplicate Entries and that's based on looking at the Original Data Frame and finding out, where there's Repeats assuming all the Values in one is found in another
- This is pretty impressive that we went through ~780,000 Rows in 1.3 Seconds and found all these Duplicates that quick
- Now, Remember, when we were were Dropping these Duplicates, we didn't specify a `subset` or a Column in the `.drop_duplicates()` Method
    - We are going to go further and Select certain Columns from the Data Frame to then, use those Columns specifically to Filter for, if there's any Duplicates
    - Based on what we know from the Data Set, let's Filter based on two things
        1. The `job_title` Column, which is very much unique. And it's so Unique that we can Filter on
            - But we don't want to Filter out all the Job Titles that are identical, just because we match on one Title. We run the risk of Removing all the Data Engineers/Analysts if we do this. So we will need something else to pivot off of, so that doesn't happen
        2. The `company_name` Column, will be our other Data Point
            - This way, we are dropping repetitive Jobs, from the same Company

In [23]:
# Still using the Uniqe Data Frame Created above, Run the Drop Duplicates Method on it
# This time, specifying the Subset. The Subset, is a List of the Column Titles we want to Filter by
df_unique = df_unique.drop_duplicates(subset=['job_title', 'company_name'])

# Get the Length of df_filled
print('Length of Original df:       ', len(df_filled))
# Get the Length of df_unique
print('Length of drop duplicates df:', len(df_unique))
# Subtract the lengths from each other to get the Total Rows Dropped
print('Rows Droppped:               ', len(df_filled)-len(df_unique))

Length of Original df:        785741
Length of drop duplicates df: 508042
Rows Droppped:                277699


### Closing Remarks
___
- From the above Output, we Dropped 277,699 Repetitive Job Postings from different Companies
- There's a lot of Companies in this Dataset that will spam Job Boards with Repetitive Jobs
- This is a good way of going through and cleaning up the Data when necessary